### ISSUE TITLE
gatsby-source-contentful: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by marcneander** [gatsbyjs/gatsby#11364]
Not sure if this is a bug or not. And unfortunately I don't really have much time this week to investigate. But I thought I might as well put this up here.

I'm using the `gatsby-source-contentful` plugin. In my contentful content model I'm having two content types, both with the new Richtext editor. Problems occurs when I have entry links to each other.

To be more clear. Say I have two pages, home and about. If I link from home -> about and from about -> home inside the Richtext editor I get this error.

Like I said before. Not sure if this is a bug.

Here's a callstack https://pastebin.com/fD676e6j

### Environment (if relevant)
```
  System:
    OS: macOS 10.14.3
    CPU: (4) x64 Intel(R) Core(TM) i5-8210Y CPU @ 1.60GHz
    Shell: 3.0.0 - /usr/local/bin/fish
  Binaries:
    Node: 10.15.0 - /var/folders/vh/cx86q1vs2xz5rxlf8b8cny2r0000gn/T/yarn--1548706327929-0.8647198662198463/node
    Yarn: 1.13.0 - /var/folders/vh/cx86q1vs2xz5rxlf8b8cny2r0000gn/T/yarn--1548706327929-0.8647198662198463/yarn
    npm: 6.4.1 - ~/.nvm/versions/node/v10.15.0/bin/npm
  Languages:
    Python: 2.7.10 - /usr/bin/python
  Browsers:
    Chrome: 71.0.3578.98
    Safari: 12.0.3
  npmPackages:
    gatsby: 2.0.98 => 2.0.98 
    gatsby-plugin-layout: 1.0.11 => 1.0.11 
    gatsby-plugin-offline: 2.0.21 => 2.0.21 
    gatsby-plugin-react-helmet: 3.0.5 => 3.0.5 
    gatsby-plugin-styled-components: 3.0.4 => 3.0.4 
    gatsby-source-contentful: 2.0.26 => 2.0.26 
    gatsby-transformer-remark: 2.2.0 => 2.2.0 
```

`gatsby-config.js`:
```
plugins: [
        {
            resolve: 'gatsby-plugin-styled-components',
            options: {
                displayName: false
            }
        },
        'gatsby-plugin-react-helmet',
        {
            resolve: 'gatsby-plugin-layout',
            options: {
                component: require.resolve('./src/components/Layout')
            }
        },
        {
            resolve: 'gatsby-source-contentful',
            options: {
                spaceId: '<spaceID>',
                accessToken: process.env.CONTENTFUL_ACCESS_TOKEN
            }
        },
        {
            resolve: '@contentful/gatsby-transformer-contentful-richtext',
            options: contentfulRenderOptions
        },
        'gatsby-plugin-offline'
    ]
```

**Comment by karenhou** [comment]
> When using reference fields, be aware that this source plugin will automatically create the reverse reference. You do not need to create references on both content types. For simplicity, it is easier to put the reference field on the child in child/parent relationships.

[https://www.gatsbyjs.org/packages/gatsby-source-contentful/?=contentful](url)

Not sure if this would help, but try to disconnect one of your reference.

**Comment by marcneander** [comment]
https://discovery.contentful.com/entries/by-content-type?delivery_access_token=809074d7507cf609df4dd9210903910a91c6f35d3ddc7ecbc2cb2cd4d5a260c4&preview=false&preview_access_token=&space_id=7b9jienzv022

https://github.com/marcneander/gatsby-contentful-bug

**Comment by simoneb** [comment]
I've been looking into this issue because we're also experiencing it. I'm not very familiar with the contentful source plugin codebase but the first impression is that the logic in place to avoid circular references only accounts for field types which are not richText. The latter is a lot more complex to analyze because it can contain a "link" to another entry in deeply nested places. I _think_ that the logic to handle this should live in the `buildForeignReferenceMap` function in the `gatsby-node.js` module, but I'm not sure about the feasibility of recursively checking the complex data structure of richText fields.

At the current time, two entries cross-referencing each other via hyperlinks inside a richText field lead to infinite recursion.

**Comment by dkhuntrods** [comment]
@sidharthachatterjee Thanks for your work on this. Would just like to add my upvote to this issue - we've encountered the exact same scenario, and I would second @simoneb's comments about this being a very valid use-case and @marcneander's repro an accurate reduced test-case of our situation. We've got a very urgent situation on our hands at the moment - do you think it likely you'll be looking at this soon? Otherwise we will look at a PR of some sort. Thanks again.

**Comment by simoneb** [comment]
Hi @marcneander, given the lack of support around here we've worked around the issue by forking the gastby repo, changing the `gatsby-source-contentful` plugin and publishing our own version on npm.

Because it was really hard to figure out how to handle rich text nodes with links to other contentful entries, we've come up with a workaround which consists in the plugin accepting an array of whitelist field names to include when processing rich text field nodes. All other fields are omitted. For the lack of a better and generic solution, this basically shifts the responsibility of which fields won't cause circular references on the shoulders of developers. 

For instance, we use the `gatsby-transformer-contentful-richtext` plugin to process rich text fields and transform hyperlinks to entries into plain `<a>` tags, in a similar way as shown in the docs:

```
[INLINES.ASSET_HYPERLINK]: node => {
  return `<img src="${
    node.data.target.fields.file['en-US'].url
  }"/>`
}
```

So, I configure our custom `gatsby-source-contentful` plugin with the option `richTextFieldEntryHyperlinkWhitelist: ['file']`, because it's the only field we use and we know it won't cause circular references.

You can find our fork [here](https://github.com/nearform/gatsby/tree/master/packages/gatsby-source-contentful). The plugin is published as [@nearform/gatsby-source-contentful](https://www.npmjs.com/package/@nearform/gatsby-source-contentful).

Hope it helps!

**Comment by snide** [comment]
@simoneb TY for your fork. This fixed the issue for me as well. It's pretty easy to get into this kind of reference, so although it's not an optimal fix I really appreciate you publishing your fork.

**Comment by simoneb** [comment]
@snide you're welcome!

**Comment by toby-bushell** [comment]
> @sidharthachatterjee Thanks for your work on this. Would just like to add my upvote to this issue - we've encountered the exact same scenario, and I would second @simoneb's comments about this being a very valid use-case and @marcneander's repro an accurate reduced test-case of our situation. We've got a very urgent situation on our hands at the moment - do you think it likely you'll be looking at this soon? Otherwise we will look at a PR of some sort. Thanks again.

I'd also like to upvote this, this is currently making internal links impossible to use in the RichText. If the reference contains any link back to the content at any level this causes the stack size explosion, which is impossible to manage or restrict.

**Comment by simoneb** [comment]
@toby-bushell as unappealing a solution as it may sound, have you tried the fork I mentioned [here](https://github.com/gatsbyjs/gatsby/issues/11364#issuecomment-462038897)? That's the solution we are applying for now.

**Comment by toby-bushell** [comment]
Hi @simoneb, am I right in thinking that the workaround is to skip trying to resolve internal links? And explicitly list the embed types to resolve in Rich Text?

**Comment by simoneb** [comment]
@toby-bushell not entirely. If you list only the fields that you need when processing your rich text, they will be resolved. The rest you will not need can safely not include in the whitelist so it doesn't cause issues.

**Comment by toby-bushell** [comment]
@simoneb thanks for this and the fork, It looks like a good failsafe option, that we may well use. 

We do however want to be able to link to other references safely within RichText and are keen on a permanent solution, as this is a very big feature of contentful. Even an understanding of the roadmap for this so we can make a call on how much time to dedicate to alternatives.

**Comment by marcneander** [comment]
Spooky quiet indeed. But not ready to be closed

**Comment by olapersson** [comment]
Having the same issue, also found a mention of the same problem here https://github.com/contentful/rich-text/issues/53 (issue was closed but it seems to me that the problem is still present).

Cross referencing entries is a quite natural thing (inline link from one page to another for example, much better to be able to use a reference rather than a hardcoded string)

As a temporary workaround I used the forked version and queried `body { body }` on the node and then parsed the json response (`richTextRenderer` being my wrapper function that maps embedded entries, images etc.)
```
      {richTextRenderer({json: JSON.parse(data.body.body) })}
```

**Comment by Khaledgarbaya** [comment]
@simoneb it can be a good starting point for a fix and we can work together on it 
there

**Comment by jessepinho** [comment]
@Khaledgarbaya I think a better solution would be to fix the underlying problem — that `gatsby-source-contentful` doesn't realize that it's already loaded a node into its internal node graph (not sure I'm using the correct terminology here), and keeps loading it in an infinite loop due to the cross-references. Probably a better solution would be for it to keep track of what it's already loaded, and point any references to the already-loaded node. Does that make sense? Kinda hard to explain by text.

**Comment by Khaledgarbaya** [comment]
This is mainly happening because there is a circular reference and somewhere in the SDK or the plugin it's not clear when to stop when resolving references.
@marcneander can you please share the json of your content types to use it as a test?

you can do that by calling this on the `contentful-cli`

### Steps

1.
```
npm install -g contentful-cli
```

2.
```
contentful login
```
and follow the instruction

3.
```
contentful space generate migration -s <space_id> -e master
```

you will get a js file that defines the content types structure that you can share it with me.

Cheers,
Khaled

**Comment by marcneander** [comment]
Sorry for being slow... Here is it [7b9jienzv022-master-1557232700604.js.zip](https://github.com/gatsbyjs/gatsby/files/3152624/7b9jienzv022-master-1557232700604.js.zip)

**Comment by xenophilicibex** [comment]
Just wanted to share that my team is also struggling with resolving this issue. Will post more details soon.

**Comment by jessepinho** [comment]
FYI, I'm working on a fix to this based on @simoneb's great work. I want to PR it, but I want to tweak the options a bit first so that you also have an option of excluding fields (rather than specifying all the fields to include).

**Comment by jessepinho** [comment]
Oh sorry, the branch is here: https://github.com/gatsbyjs/gatsby/compare/master...jessepinho:topics/rich-text-field-filter

**Comment by thomaskrantz** [comment]
Just wanted to chime in that we're also experiencing the same issue, and I have tried using @jessepinho's PR but for some reason that didn't have any effect. Not sure how I can help though.

**Comment by jessepinho** [comment]
@thomaskrantz did you also pass a `richText.includeEntryFields` option? (I noticed btw that the documentation I included had a typo of `includeEntryList` when it should have been `includeEntryFields`)

**Comment by cameron-martin** [comment]
@Khaledgarbaya does your assignment of this mean that you're in the process of fixing it? I was going to take a look at this, but if you're looking at it then I'll hold off.

**Comment by Khaledgarbaya** [comment]
@cameron-martin please feel free to take a look at it

**Comment by dwikberg** [comment]
We are also experiencing this issue. I will Fedex a bottle of Dom Perignon champagne to whoever solves this!!

**Comment by simoneb** [comment]
@dwikberg If [our workaround](https://github.com/gatsbyjs/gatsby/issues/11364#issuecomment-462038897) works for you as it did for us you can post the bottle to my address ;)

**Comment by dwikberg** [comment]
> @dwikberg If [our workaround](https://github.com/gatsbyjs/gatsby/issues/11364#issuecomment-462038897) works for you as it did for us you can post the bottle to my address ;)

We need a more permanent solution :(

**Comment by simoneb** [comment]
Considering that this has been open for 6 months now, I'd rather stick with the workaround, we've been running that in production since then and based on my experience it covers most common scenarios. I'm ok with a bottle of Pinot Noir as well as this is not a permanent solution.

**Comment by dwikberg** [comment]
> Considering that this has been open for 6 months now, I'd rather stick with the workaround, we've been running that in production since then and based on my experience it covers most common scenarios. I'm ok with a bottle of Pinot Noir as well as this is not a permanent solution.

I think we are using a newer version with gatsby where the workaround does not cover some of the scenarios. Our build keeps on crashing :) But A for effort :)

**Comment by jayhostan** [comment]
Just run into the same thing. Is @simoneb 's workaround still the best temp. option?

**Comment by jessepinho** [comment]
@jayhostan I think my PR (#14612) provides a decent temp solution. 

However, a longer-term solution (which I discussed offline with @Khaledgarbaya) would be to properly match Contentful’s real GraphQL API. In [their docs](https://www.contentful.com/developers/docs/references/graphql/#/reference/schema-generation/rich-text), they describe a `links` property that lists all linked entries. Then you could simply control which fields of which content types you want to pull. I believe this would resolve the infinite-recursion issue, but I haven’t explored it enough yet to know for sure.

**Comment by kompounder** [comment]
Hi - is there a solution to this issue? It seems to severely limit Contentful's functionality with Gatsby. Any update would be much appreciated!

**Comment by pvdz** [comment]
We've landed a fix for the stack overflow caused by circular dependencies. That fix should resolve this issue. If not, please reopen with a new (minimal) repro.

Thanks everyone for helping out and for your patience.

**Comment by kompounder** [comment]
Thank you very much for taking a look at this.
Unfortunately Gatsby 2.18.5 continues to give me the same error on Contentful rich texts ("maximum call stack size exceeded") whenever a richtext field has an embedded reference to a content type that includes a richtext field.
Will try to submit a minimal repro.

**Comment by pvdz** [comment]
@kompounder  oh :( If you can create a repro please let me know. Note that I suspect it's not actually related to contentful (though it might be). A repro that doesn't use contentful would be even better and is probably easier to create because you can do it locally.

**Comment by richhiggins** [comment]
this is still broken for me using the latest 2.18.6. same error message, looks to be caused by using a link type of entry within a Rich Text editor.

**Comment by richhiggins** [comment]
interestingly I have another project where I've used the rich text editor in a referenced type (I called it `paragraph`) rather than a field within my `page` type. my page entries include multiple paragraph referenced entries.

![image](https://user-images.githubusercontent.com/3781432/70246563-dd4f6c80-176f-11ea-9266-00e8eb7fdb88.png)

in this scenario linking to the `page` entry does not cause the stack bug, I'm presuming because the source type is `paragraph` rather than `page`.

Would that count as a (content model based) permanent fix @dwikberg 🍾?

**Comment by jessepinho** [comment]
@richhiggins it will happen if your linked entry links back to the current entry, thus creating an infinite loop

**Comment by pvdz** [comment]
It would be great if you could create a minimal repro with tihs, or at least a repo that I can just clone and build to repro the case if nothing else.

I'll try to revisit this on Monday. Thanks for the input!

**Comment by pvdz** [comment]
I'm removing the "not stale" label until somebody can come up with a repro. If nothing else this issue will be closed by the bot in about 30 days.

If somebody has a repro I can check out and confirm, please post it here. Thanks!

**Comment by cameron-martin** [comment]
There is a repro in the form of a functional test in this PR: https://github.com/gatsbyjs/gatsby/pull/15358

It's not isolated from contentful, but it should demonstrate the issue still.

**Comment by pvdz** [comment]
Ok. Can repro. On it.

**Comment by pvdz** [comment]
@cameron-martin Thank you. I was able to add a repro to the tests that caught this stack overflow and able to prevent the stack overflow. Please report back after testing with #20039 .

**Comment by pvdz** [comment]
@cameron-martin This repro was fixed in #20039 which has been published. Please check whether this works for you.

I'm going to close this for now. If anyone has related problems feel free to reopen. Or a new issue and tag me.

Thanks!

**Comment by jessepinho** [comment]
🚨 *I've created a minimal reproduction of this issue!* 🚨 

This appears to _still_ be a problem on the latest version of Gatsby, even after #20039 was merged.

Here is the repo: https://github.com/jessepinho/gatsby-contentful-infinite-loop-repro

And a demonstration GIF:
![screencap](https://user-images.githubusercontent.com/1121544/72349799-7e213800-36dd-11ea-99df-98459c97938b.gif)

@pvdz could you please reopen this issue? (I don't have rights to.)

**Comment by pvdz** [comment]
Great, thank you!

**Comment by pvdz** [comment]
Confirmed. Looking into a fix now.

**Comment by pvdz** [comment]
The fix (which at least resolves your repro) is in #20674

Follow that to see when it gets merged and to see what version it gets released in.

**Comment by jessepinho** [comment]
@pvdz WOW that was fast, thank you! When I get a chance, I'll try it with my real-life repo and see if it fixes the issue.

**Comment by pvdz** [comment]
With #20674 being merged this repro has been fixed. It should be published on the next patch bump (should be posted on PR).

Please feel free to reopen if you're still having issues after that.

**Comment by ma-pe** [comment]
I can confirm that the fix (released with v 2.7.8) solves my problem 👍 

Thanks a lot!

**Comment by jessepinho** [comment]
I just upgraded and am still experiencing the problem in my own repo. @ma-pe what library do you have v2.7.8 of? There is no `gatsby-source-contentful` with that version...?

**Comment by pvdz** [comment]
@jessepinho are you saying the repro you posted is still having this problem with the above fix in place? I couldn't repro it anymore when it landed. But please let me know if you can repro it when all deps are up to date (it may help to just delete the `node_modules` entirely and make sure the package.json is up to date).

**Comment by vpreponen** [comment]
I assumed he meant 2.1.78. I also continued to experience the problem with v2.1.78 so I continued to use a work-around with legacy versions.

**Comment by pvdz** [comment]
@vpreponen The current published version is 2.1.85, can you check if you're still experiencing it with that version? (Is there anything blocking you from upgrading the patch bump?)

**Comment by jessepinho** [comment]
@pvdz not the repro I posted, but a separate repo that I was trying to reproduce with that minimal repro I posted. Apparently it was a little too minimal 😆 Will look into this further. And yeah I used `v2.1.85`, and upgrade Gatsby to `2.19.16`.

**Comment by pvdz** [comment]
@jessepinho ok, if you manage to get another repro repo or test case please share it and I'll have a look and try to fix it :)

**Comment by vpreponen** [comment]
Deleting lock files and node_modules plus updating to gatsby v2.19.16 and gatsby-source-contentful v2.1.85 fixes the issue for me.

**Comment by icavalheiro** [comment]
I can confirm this happens on Nuxt.JS as well. So might not be an issue in Gatsby?

**Comment by inblock-io** [comment]
> @jessepinho ok, if you manage to get another repro repo or test case please share it and I'll have a look and try to fix it :)

I have one here: https://github.com/gatsbyjs/gatsby/issues/27515

**Comment by StyvekDrtic** [comment]
im still having problems...
